### PR TITLE
BYOC/guest cluster scoped aws client - resources implementation

### PR DESCRIPTION
**Pull request by TheoBrigitte**
Towards: https://github.com/giantswarm/giantswarm/issues/2947

Resources are now using guest cluster scoped aws client (provided by https://github.com/giantswarm/aws-operator/pull/916).
Updated resources:
* kmsKeyResource
* s3BucketResource
* s3BucketObjectResource
* loadBalancerResource
* ebsVolumeResource
* cloudformationResource
* endpointsResource

All services we use to communicate with aws (`awsService`, `cloudConfig`, `ebsService`, `cloudFormationService`) are now created in the `initCtxFunc` function
https://github.com/giantswarm/aws-operator/blob/byoc/guest-cluster-scoped-aws-client-resources/service/controller/v10/cluster_resource_set.go#L317

Major changes is adoption of aws services interfaces for `client/aws` and aws mocks.
e.g. https://godoc.org/github.com/aws/aws-sdk-go/service/ec2/ec2iface

### CODE REVIEW

File: service/controller/v10/adapter/mocks.go [hunk 1]
```diff
@@ -40,9 +46,11 @@ func (e *EC2ClientMock) DescribeSecurityGroups(input *ec2.DescribeSecurityGroups
 	return nil, fmt.Errorf("security group not found")
 }
 
-func (e *EC2ClientMock) DescribeSubnets(input *ec2.DescribeSubnetsInput) (*ec2.DescribeSubnetsOutput, error) {
-	if e.subnetID == "" {
-		e.subnetID = "subnet-1234"
+func (e EC2ClientMock) DescribeSubnets(input *ec2.DescribeSubnetsInput) (*ec2.DescribeSubnetsOutput, error) {
```
Comment (TheoBrigitte) [inline service/controller/v10/adapter/mocks.go]: Changing methods to non-pointer receiver is required to comply with ec2iface.EC2API interface.

**Comment by TheoBrigitte** [comment]
I did got started on this before the freeze. I all have to port all of this to `v11` ...

**Comment by xh3b4sd** [comment]
Yeah, lets talk when you find a cool way of doing this. I heard something about using patch files and such. Maybe you find a cool way which we can use to add it to our tooling so we can execute one `opsctl` command in the future and have the move sorted for us.

File: client/aws/aws.go [hunk 2]
```diff
@@ -32,12 +37,12 @@ type Config struct {
...
-	ELB            *elb.ELB
-	IAM            *iam.IAM
-	KMS            *kms.KMS
+	EC2            ec2iface.EC2API
+	ELB            elbiface.ELBAPI
+	IAM            iamiface.IAMAPI
+	KMS            kmsiface.KMSAPI
 	Route53        *route53.Route53
-	S3             *s3.S3
+	S3             s3iface.S3API
```
Comment (kopiczko) [inline client/aws/aws.go]: Nice you found those interfaces.

File: service/controller/v10/adapter/mocks.go [hunk 3]
```diff
@@ -7,12 +7,18 @@ import (
...
 	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/aws/aws-sdk-go/service/elb/elbiface"
 	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/iam/iamiface"
 	"github.com/aws/aws-sdk-go/service/kms"
+	"github.com/aws/aws-sdk-go/service/kms/kmsiface"
 )
 
 type EC2ClientMock struct {
+	ec2iface.EC2API
```
Comment (kopiczko) [inline service/controller/v10/adapter/mocks.go]: Why is that line needed?

File: service/controller/v10/context/awsservice/context.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package awsservice
```
Comment (kopiczko) [inline service/controller/v10/context/awsservice/context.go]: Instead of all this context packages, can't we have one struct which holds all of those and store it inside the context?

File: service/controller/v10/adapter/mocks.go (hunk 1 above)
Comment (TheoBrigitte) [reply service/controller/v10/adapter/mocks.go]: Otherwise we run in this issue
```
# github.com/giantswarm/aws-operator/service/controller/v10/resource/cloudformation
service/controller/v10/resource/cloudformation/create_test.go:106:6: cannot use adapter.EC2ClientMock literal (type adapter.EC2ClientMock) as type ec2iface.EC2API in field value:
        adapter.EC2ClientMock does not implement ec2iface.EC2API (DescribeSubnets method has pointer receiver)
```

File: service/controller/v10/adapter/mocks.go (hunk 3 above)
Comment (TheoBrigitte) [reply service/controller/v10/adapter/mocks.go]: This is needed to have `EC2ClientMock` fulfills the `ec2iface.EC2API` interface so we can later on assign `EC2ClientMock` to aws.Clients.EC2 which accept a `ec2iface.EC2API` interface.
https://github.com/giantswarm/aws-operator/pull/933/files#diff-1044bb920626db9f96ba5e53d1c4295cR103

File: service/controller/v10/context/awsservice/context.go (hunk 4 above)
Comment (TheoBrigitte) [reply service/controller/v10/context/awsservice/context.go]: Yes good point, I will refactor that as a single `context` package and storing all those together.

File: service/controller/v10/adapter/mocks.go (hunk 3 above)
Comment (kopiczko) [reply service/controller/v10/adapter/mocks.go]: In order to fulfill the interface you don't need to embed it. I think that's also the reason you had to remove pointer receivers.

File: service/controller/v10/adapter/mocks.go (hunk 1 above)
Comment (kopiczko) [reply service/controller/v10/adapter/mocks.go]: What if you pass `& adapter.EC2ClientMock`? Why `adapter` doesn't hold the pointer?

File: service/controller/v10/adapter/mocks.go (hunk 3 above)
Comment (TheoBrigitte) [reply service/controller/v10/adapter/mocks.go]: To fulfill the Interface I would otherwise have to implement all its methods. Embedding the interface allow to fulfill the interface without having to implements all its methods.

The downside is that calling an method which is not implemented will result in panic, but since we use those mock in our tests we know what we are doing here.

Comment (kopiczko) [reply service/controller/v10/adapter/mocks.go]: I didn't know that. It's interesting.

But you should be able to leave the pointer receivers: https://play.golang.org/p/auB-Dm5cD0a

Comment (TheoBrigitte) [reply service/controller/v10/adapter/mocks.go]: Yes I actually tried this and it worked.

**Comment by TheoBrigitte** [comment]
I ported changes from `v10` to `v11` and address some changes.

In order to port controller specific changes the most efficient way I found was a combination of `git diff` and renaming import path.
```
git diff  <a> <b> service/controller/v10 > controller.patch
replace "service/controller/v10" "service/controller/v11" controller.patch 
git apply controller.patch
```

credits goes to kopiczko for the replace function

For the records I used `git diff <a> <b> ":\!service/controller/v10" ":\!vendor" ":\!Gopkg*"` to port other changes with some manual adjustment and a `dep ensure` in the end.

**Comment by xh3b4sd** [comment]
Can you please add some docs for this to https://github.com/giantswarm/giantswarm/wiki/Releases:-Creating-New-VersionBundle? Maybe add a trouble shooting section or best practises or whatever fits best for you and explain your case. When we collect these cases it is easier to go through the list in order to pick the biggest pain point to automate via tooling.

**Comment by TheoBrigitte** [comment]
Documented in https://github.com/giantswarm/giantswarm/wiki/Releases:-Creating-New-VersionBundle#port-changes-from-one-version-to-the-next

File: service/controller/v10/adapter/mocks.go (hunk 1 above)
Comment (TheoBrigitte) [reply service/controller/v10/adapter/mocks.go]: done

**Comment by TheoBrigitte** [comment]
Merging in parent branch, as separate PRs does not make sense anymore.
